### PR TITLE
feat(storage): adds buying storage capability

### PR DESCRIPTION
**Pull request by jurajpiar**
Resolves [Buy Storage (Main - Hash](https://rsklabs.atlassian.net/browse/RMKT-229)

### CODE REVIEW

File: src/components/molecules/storage/buy/RifSelect.tsx [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+import React from 'react'
+
+type OptionType = JSX.Element | string
+
+export type RifSelectProps<T extends OptionType> = SelectProps & {
+    id: string
+    options: T[] | undefined
+}
+
+function RifSelect<T extends OptionType>({
```
Comment (farinalvaro) [inline src/components/molecules/storage/buy/RifSelect.tsx:11]: Looks good! Could we extract it outside of storage folder? Also may be create an issue to extract to `rif-ui`

File: src/components/organisms/RifCard.tsx [hunk 2]
```diff
@@ -0,0 +1,74 @@
+import React, { FC } from 'react'
+import {
+  Card, CardActions, CardContent, CardProps, makeStyles, Theme,
+} from '@material-ui/core'
+import GridColumn from 'components/atoms/GridColumn'
+import GridItem from 'components/atoms/GridItem'
+
+const useStylesRifCard = makeStyles((theme: Theme) => ({
+  card: {
+    background: '#F8F7F7',
```
Comment (farinalvaro) [inline src/components/organisms/RifCard.tsx:10]: Try to use [the colors of our palette ](https://github.com/rsksmart/rif-ui/blob/master/src/theme/theme.ts#L4) or add a new one there

File: src/components/organisms/RifCard.tsx [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+  },
+  titleContainer: {
+    padding: theme.spacing(3),
+  },
+  contentContainer: {
+    width: '395px',
+    minHeight: '276px',
+    margin: '50px',
+    marginTop: '10px',
+    background: '#FFF',
```
Comment (farinalvaro) [inline src/components/organisms/RifCard.tsx:24]: Try to use [the colors of our palette ](https://github.com/rsksmart/rif-ui/blob/master/src/theme/theme.ts#L4) or add a new one there

**Review by farinalvaro — COMMENTED** [review]
Looking good! My review is still in progress but I just left a few comments

File: src/components/molecules/DropZone.tsx [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+import {
+  makeStyles, Theme,
+} from '@material-ui/core'
+import { colors } from '@rsksmart/rif-ui'
+import GridColumn from 'components/atoms/GridColumn'
+
+const useStylesDropZone = makeStyles((theme: Theme) => ({
+  root: {
+    borderRadius: theme.spacing(1.2),
+    border: '1px dashed #CFD3DA',
```
[the remark this answers is not in the document]
Comment (jurajpiar) [reply src/components/molecules/DropZone.tsx:12]: Actually I just forgot to remove this component as I'm not using it in the end.

File: src/components/molecules/storage/buy/LabelWithValue.tsx [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+export type LabelWithValueProps = Modify<GridProps,
+    {
+        label: string
+        value: string
+        labelProps?: TypographyProps
+        valueProps?: TypographyProps
+    }>
+
+const LabelWithValue: FC<LabelWithValueProps> = ({
```
[the remark this answers is not in the document]
Comment (jurajpiar) [reply src/components/molecules/storage/buy/LabelWithValue.tsx:18]: out of scope

File: src/components/molecules/storage/buy/PinEnterInfoTab.tsx [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+const isEmpty = (
+  text: string | unknown,
+): boolean => !text || !String(text).trim()
+
+const PinEnterInfoTab: FC<Props> = ({
+  name, size, unit, hash,
+}) => (
+  <>
+    <GridRow justify="center">
+      {/* .juraj - put inside classes */}
```
[the remark this answers is not in the document]
Comment (jurajpiar) [reply src/components/molecules/storage/buy/PinEnterInfoTab.tsx:25]: no point in doing that for one thing only is it

File: src/components/molecules/storage/buy/RifSelect.tsx (hunk 1 above)
Comment (jurajpiar) [reply src/components/molecules/storage/buy/RifSelect.tsx:11]: https://github.com/rsksmart/rif-marketplace-cache/issues/318

File: src/components/molecules/storage/buy/LabelWithValue.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (jurajpiar) [reply src/components/molecules/storage/buy/LabelWithValue.tsx:18]: Looks like I didn't use it in the end so removing.

**Review by jurajpiar — COMMENTED** [review]
Happy reviewing!

File: src/components/molecules/storage/buy/RifSelect.tsx (hunk 1 above)
Comment (farinalvaro) [reply src/components/molecules/storage/buy/RifSelect.tsx]: That issue was created in marketplace-cache 😄

File: src/components/molecules/storage/buy/PinUploaderTab.tsx [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+          <DropzoneArea
+            {...props}
+            classes={{ root: classes.root, text: classes.text }}
+            showPreviews
+            showPreviewsInDropzone={false}
+            useChipsForPreview
+            previewGridProps={{ container: { spacing: 1, direction: 'row' } }}
+            previewChipProps={{ classes: { root: classes.previewChip } }}
+            previewText="Selected files"
+          />
```
Comment (farinalvaro) [inline src/components/molecules/storage/buy/PinUploaderTab.tsx:53]: Nice! 👌

File: src/components/molecules/storage/buy/PlanOption.tsx [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+const PlanOption: FC<PlanOptionProps> = ({ plan, xr }) => {
+  const { currency, period, price } = plan
+  const { fiat, rate } = xr
+  return (
+    <GridRow
+      wrap="nowrap"
+      key={currency + period + price}
+      style={{
+        paddingBlockEnd: '0em',
+      }}
```
Comment (farinalvaro) [inline src/components/molecules/storage/buy/PlanOption.tsx:26]: I assume this component is going to be rendered multiple times, if that's the case I would suggest to not use inline styles. Using inline would result in duplicated css in the DOM. Reference: https://stackoverflow.com/a/57812892/1361335

File: src/components/organisms/RifCard.tsx [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+    <Card className={classes.card} variant="outlined" {...props}>
+      <GridColumn
+        alignContent="center"
+        justify="center"
+        className={classes.titleContainer}
+      >
+        {!!Header && (
+        <GridItem style={{
+          alignSelf: 'center',
+        }}
```
Comment (farinalvaro) [inline src/components/organisms/RifCard.tsx:55]: If possible, move to classes

File: src/components/molecules/storage/buy/RifSelect.tsx (hunk 1 above)
Comment (jurajpiar) [reply src/components/molecules/storage/buy/RifSelect.tsx]: lol

Comment (jurajpiar) [reply src/components/molecules/storage/buy/RifSelect.tsx]: https://github.com/rsksmart/rif-marketplace-ui/issues/382

File: src/components/organisms/storage/buy/PinningCard.tsx [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+const useActionButtonStyles = makeStyles(() => ({
+  disabled: {
+    background: colors.gray1,
+  },
+}))
+
+const PinningCard: FC<Props> = ({ dispatch }) => {
+  const actionBtnClasses = useActionButtonStyles()
+
+  const [isUpladed, setIsUploaded] = useState(false)
```
Comment (farinalvaro) [inline src/components/organisms/storage/buy/PinningCard.tsx:31]: `isUploaded`*

File: src/components/organisms/storage/buy/PinningCard.tsx [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+      payload: {
+        name, size, unit: unit || undefined, hash,
+      },
+    })
+  }
+
+  const handleUpload = async (): Promise<void> => {
+    // Upload files
+    await Promise.resolve(files)
+    // Update context
```
Comment (farinalvaro) [inline src/components/organisms/storage/buy/PinningCard.tsx:53]: Did you forget to update context here?

File: src/components/organisms/storage/buy/StoragePinTabs.tsx [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+}))
+
+export type StoragePinTabsProps = Modify<TabsProps, {
+  value: boolean
+  onChange: (event: React.ChangeEvent<{}>, value: unknown) => void
+}>
+
+const StoragePinTabs: FC<StoragePinTabsProps> = ({ ...props }) => {
+  const tabsStylesClasses = useTabsStyles()
+  const tabItemStylesClasses = useTabItemStyles()
```
Comment (farinalvaro) [inline src/components/organisms/storage/buy/StoragePinTabs.tsx:36]: 😍

File: src/components/organisms/storage/buy/StoragePinTabs.tsx [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+
+  return (
+    <Tabs
+      classes={tabsStylesClasses}
+      indicatorColor="primary"
+      textColor="primary"
+      {...props}
+    >
+      <Tab classes={tabItemStylesClasses} label="Upload file" value={false} />
+      <Divider orientation="vertical" flexItem style={{ height: '29px', width: '2px', alignSelf: 'center' }} />
```
Comment (farinalvaro) [inline src/components/organisms/storage/buy/StoragePinTabs.tsx:46]: If possible, move to classes too

File: src/components/pages/storage/buy/CheckoutContext.tsx [hunk 14]
```diff
@@ -0,0 +1,517 @@
+import React, {
```
Comment (farinalvaro) [inline src/components/pages/storage/buy/CheckoutContext.tsx:1]: I think this file is too big, could we separate into Context, reducer and actions?
Also think that would fit better under `contexts/storage/buy/checkout` folder than in `components/pages` folder

File: src/components/pages/storage/buy/StorageOffersCheckoutPage.tsx [hunk 15]
```diff
@@ -0,0 +1,281 @@
...
+      <GridColumn>
+        <GridItem>
+          {order && pinned && <StorageOrderDescription order={{ ...order, ...pinned }} />}
+        </GridItem>
+        {/* STEPPER */}
+        { renderStepper() }
+        {/* CONTENT */}
+        { renderContent() }
+      </GridColumn>
+      {renderProgressOverlay()}
```
Comment (farinalvaro) [inline src/components/pages/storage/buy/StorageOffersCheckoutPage.tsx:277]: Is it possible to extract `renderStepper`, `renderContent` and `renderProgressOverlay` to new components under `components/organisms/storage/buy`?

**Review by farinalvaro — COMMENTED** [review]
Just pointed some details but overall the amount of work of this PR is impressive 🙂

File: src/components/molecules/storage/buy/PinEnterInfoTab.tsx [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+type Props = {
+    name: TextFieldProps
+    size: TextFieldProps
+    hash: TextFieldProps
+    unit: SelectProps
+}
+
+const isEmpty = (
+  text: string | unknown,
+): boolean => !text || !String(text).trim()
```
[the remark this answers is not in the document]
Comment (jurajpiar) [reply src/components/molecules/storage/buy/PinEnterInfoTab.tsx:19]: it's only used here

[the remark this answers is not in the document]
Comment (jurajpiar) [reply src/components/molecules/storage/buy/PinEnterInfoTab.tsx:19]: but can do

File: src/components/organisms/storage/buy/PinningCard.tsx (hunk 11 above)
Comment (jurajpiar) [reply src/components/organisms/storage/buy/PinningCard.tsx:53]: nope. this is just a prep for the upload US.

File: src/components/pages/storage/buy/CheckoutContext.tsx (hunk 14 above)
Comment (jurajpiar) [reply src/components/pages/storage/buy/CheckoutContext.tsx:1]: split - yes can do
move to context - I disagree but you know that, so I'm gonna move it to the context folder.

Comment (jurajpiar) [reply src/components/pages/storage/buy/CheckoutContext.tsx:1]: btw better structure is Context, actions and interfaces. Reducer is not accurate and may confuse.

Comment (farinalvaro) [reply src/components/pages/storage/buy/CheckoutContext.tsx:1]: Agree 👍

**Comment by julianmrodri** [comment]
Works in general. There is the known issue with the `size` which is not reducing by the correct amount but we will track in separate issue.

**Review by julianmrodri — APPROVED** [review]
LGTM

File: src/components/pages/storage/buy/StorageOffersCheckoutPage.tsx (hunk 15 above)
Comment (jurajpiar) [reply src/components/pages/storage/buy/StorageOffersCheckoutPage.tsx:287]: this wouldn't add much benefit to the project in fact would unnecessarily create complexity. Don't get fooled by the return type. it still is just a function.

**Review by farinalvaro — APPROVED** [review]

**Pull request merged**